Verso/PolychromaticSite.lean

```diff
@@ -79,6 +129,89 @@ class B green;
+2. For each translate $`n + S` and each colour $`c`, define a "bad event" as the event that no
+   element of $`n + S` receives colour $`c`.
+3. The probability of a bad event is $`(1 - 1/k)^m`, which is small when $`m` is large relative
+   to $`k`.
+4. Bad events for translates $`n + S` and $`n' + S` are independent unless the translates overlap,
+   which bounds the dependency.
+
+The Local Lemma works for finitely many translations. The Rado selection principle (a topological
+compactness argument) extends this to all translations. Then straightforward analysis gives the
+bound $`m(k) \leq 3k^2`, and a more careful asymptotic analysis gives $`m(k) \leq (3 + o(1)) k \log k`.
+This latter bound is optimal up to constant factors (the true value is $`(1 + o(1)) k \log k`).
+
+The quadratic bound is formalised in Lean:
+
+```anchor exists_colouring_of_sq_le (module := Polychromatic.Existence)
+theorem exists_colouring_of_sq_le {S : Finset G} {k : ℕ} (hk : k ≠ 0) (hm : 3 * k ^ 2 ≤ #S) :
+    HasPolychromColouring (Fin k) S := by
+```
+
+We also prove the lower bound $`k \leq m(k)`:
+
+```anchor le_straussFunction_self (module := Polychromatic.LovaszFunction)
+lemma le_straussFunction_self {k : ℕ} :
+    k ≤ straussFunction k := by
+```
+
+# The Four-Three Problem
+
+The statement that every set of 4 integers admits a 3-polychromatic colouring is known as the
+*four-three problem*. The proof combines theoretical reductions with exhaustive computational
+verification.
+
+## Theoretical Reductions
+
+Without loss of generality, we may assume:
+1. The minimum element of $`S` is 0 (by translation invariance).
+2. The elements are coprime (scaling preserves the polychromatic number in torsion-free groups).
+3. The elements satisfy certain ordering conditions.
+
+These reductions allow us to focus on quadruples of the form $`\{0, a, b, c\}` with
+$`0 < a < b < c`, $`a + b \leq c`, and $`\gcd(a, b, c) = 1`.
+
+## Computational Verification
+
+For quadruples with $`c < 289`, the proof is completed by computational verification. The approach
+uses:
+1. *Periodic colourings*: A colouring with period $`q$ is represented as a function
+   $`\mathbb{Z} \to \text{Fin } 3$ that repeats every $`q` integers.
+2. *Bit vector encoding*: Each period-$`q` colouring is encoded as three bit vectors (one per
+   colour), enabling efficient verification that every translate hits all colours.
+3. *Residue shortcuts*: Certain cases can be quickly discharged using simple modular arithmetic
+   (e.g., if $`a \equiv 1`, $`b \equiv 2 \pmod 3`).
+
+The C++ code in the `Generation` directory produces colouring witnesses for the vast majority of the
+900,000 sets, but for the cases where a large ($`\geq 30`) period $`q` is necessary, this algorithm
+is too slow. For these, we use Z3Py to find a colouring instead. These are then verified in Lean
+using metaprogramming to construct proof terms.
```

Comment:
This is not a very clear explanation, make it easier to understand. You could also switch this paragraph with the previous, to reflect the logical structure. That is, we first try the C++ code for all colourings. For the ones where a search up to period 30 fails, we try Z3Py. Then, Lean verifies them all using the three steps. Note that the periodic colouring is actually represented y a function Z/qZ -> Fin 3